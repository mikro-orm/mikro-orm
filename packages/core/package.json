{
  "name": "@mikro-orm/core",
  "version": "6.1.12",
  "description": "TypeScript ORM for Node.js based on Data Mapper, Unit of Work and Identity Map patterns. Supports MongoDB, MySQL, PostgreSQL and SQLite databases as well as usage with vanilla JavaScript.",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "typings": "dist/index.d.ts",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.mjs"
      },
      "require": "./dist/index.js"
    }
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com/mikro-orm/mikro-orm.git"
  },
  "funding": "https://github.com/sponsors/b4nan",
  "keywords": [
    "orm",
    "mongo",
    "mongodb",
    "mysql",
    "mariadb",
    "postgresql",
    "sqlite",
    "sqlite3",
    "ts",
    "typescript",
    "js",
    "javascript",
    "entity",
    "ddd",
    "mikro-orm",
    "unit-of-work",
    "data-mapper",
    "identity-map"
  ],
  "author": "Martin AdÃ¡mek",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/mikro-orm/mikro-orm/issues"
  },
  "homepage": "https://mikro-orm.io",
  "engines": {
    "node": ">= 18.12.0"
  },
  "scripts": {
    "build": "yarn clean && yarn compile && yarn copy && yarn run -T gen-esm-wrapper dist/index.js dist/index.mjs",
    "clean": "yarn run -T rimraf ./dist",
    "compile": "yarn run -T tsc -p tsconfig.build.json",
    "copy": "node ../../scripts/copy.mjs"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "dataloader": "2.2.2",
    "dotenv": "16.4.5",
    "esprima": "4.0.1",
    "fs-extra": "11.2.0",
    "globby": "11.1.0",
    "mikro-orm": "6.1.12",
    "reflect-metadata": "0.2.2"
  }
}
