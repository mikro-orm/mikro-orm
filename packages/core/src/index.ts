/**
 * @packageDocumentation
 * @module core
 */
/* istanbul ignore file */
export {
  Constructor, Dictionary, PrimaryKeyType, PrimaryKeyProp, Primary, IPrimaryKey, FilterQuery, IWrappedEntity, EntityName, EntityData, Highlighter,
  AnyEntity, EntityProperty, EntityMetadata, QBFilterQuery, PopulateOptions, Populate, Loaded, New, LoadedReference, LoadedCollection,
  GetRepository, EntityRepositoryType, MigrationObject, DeepPartial, PrimaryProperty, Cast, IsUnknown, EntityDictionary, EntityDTO, PlainObject,
} from './typings';
export * from './enums';
export * from './errors';
export * from './exceptions';
export * from './MikroORM';
export * from './entity';
export * from './events';
export * from './EntityManager';
export * from './unit-of-work';
export * from './utils';
export * from './hydration';
export * from './drivers';
export * from './connections';
export * from './platforms';
export * from './types';
export * from './naming-strategy';
export * from './metadata';
export * from './cache';
export * from './decorators';
