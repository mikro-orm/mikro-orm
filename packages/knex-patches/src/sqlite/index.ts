export * from './SqliteKnexDialect';
export * from './SqliteTableCompiler';
