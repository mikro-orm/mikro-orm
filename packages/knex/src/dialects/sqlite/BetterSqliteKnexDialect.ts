import { SqliteTableCompiler } from './SqliteTableCompiler';
import { MonkeyPatchable } from '../../MonkeyPatchable';

export class BetterSqliteKnexDialect extends MonkeyPatchable.BetterSqlite3Dialect {

  _driver() {
    return require('better-sqlite3');
  }

  tableCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (SqliteTableCompiler as any)(this, ...arguments);
  }

}
