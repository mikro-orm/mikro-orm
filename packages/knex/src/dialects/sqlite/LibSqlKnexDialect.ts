import { SqliteTableCompiler } from './SqliteTableCompiler';
import { MonkeyPatchable } from '../../MonkeyPatchable';

export class LibSqlKnexDialect extends MonkeyPatchable.BetterSqlite3Dialect {

  get driverName() {
    return 'libsql';
  }

  _driver() {
    return require('libsql');
  }

  tableCompiler() {
    // eslint-disable-next-line prefer-rest-params
    return new (SqliteTableCompiler as any)(this, ...arguments);
  }

}
