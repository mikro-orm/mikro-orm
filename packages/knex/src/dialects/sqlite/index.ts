export * from './BaseSqliteConnection';
export * from './BaseSqlitePlatform';
export * from './BaseSqliteSchemaHelper';
export * from './SqliteTableCompiler';
export * from './BetterSqliteKnexDialect';
export * from './LibSqlKnexDialect';
export * from './SqliteKnexDialect';
