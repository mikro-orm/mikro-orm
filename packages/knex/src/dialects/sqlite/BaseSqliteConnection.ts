import { dirname } from 'node:path';
import { CompiledQuery } from 'kysely';
import { Utils } from '@mikro-orm/core';
import { AbstractSqlConnection } from '../../AbstractSqlConnection.js';

export abstract class BaseSqliteConnection extends AbstractSqlConnection {

  override async connect(simple = false): Promise<void> {
    await super.connect();

    if (simple) {
      return;
    }

    const dbName = this.config.get('dbName');

    if (dbName && dbName !== ':memory:') {
      Utils.ensureDir(dirname(this.config.get('dbName')!));
    }

    await this.client.executeQuery(CompiledQuery.raw('pragma foreign_keys = on'));
  }

  override getClientUrl(): string {
    return '';
  }

}
