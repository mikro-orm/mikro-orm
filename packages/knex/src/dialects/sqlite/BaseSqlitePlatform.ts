import { Utils } from '@mikro-orm/core';
import { AbstractSqlPlatform } from '../../AbstractSqlPlatform';

export abstract class BaseSqlitePlatform extends AbstractSqlPlatform {

  override usesDefaultKeyword(): boolean {
    return false;
  }

  override usesReturningStatement(): boolean {
    return true;
  }

  override getCurrentTimestampSQL(length: number): string {
    return super.getCurrentTimestampSQL(0);
  }

  override getDateTimeTypeDeclarationSQL(column: { length: number }): string {
    return 'datetime';
  }

  override getEnumTypeDeclarationSQL(column: { items?: unknown[]; fieldNames: string[]; length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    if (column.items?.every(item => Utils.isString(item))) {
      return 'text';
    }

    /* istanbul ignore next */
    return this.getTinyIntTypeDeclarationSQL(column);
  }

  override getTinyIntTypeDeclarationSQL(column: { length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    return this.getIntegerTypeDeclarationSQL(column);
  }

  override getSmallIntTypeDeclarationSQL(column: { length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    return this.getIntegerTypeDeclarationSQL(column);
  }

  override getIntegerTypeDeclarationSQL(column: { length?: number; unsigned?: boolean; autoincrement?: boolean }): string {
    return 'integer';
  }

  override getFloatDeclarationSQL(): string {
    return 'real';
  }

  override getBooleanTypeDeclarationSQL(): string {
    return 'integer';
  }

  override getVarcharTypeDeclarationSQL(column: { length?: number }): string {
    return 'text';
  }

  override convertsJsonAutomatically(): boolean {
    return false;
  }

  override allowsComparingTuples() {
    return false;
  }

  /**
   * This is used to narrow the value of Date properties as they will be stored as timestamps in sqlite.
   * We use this method to convert Dates to timestamps when computing the changeset, so we have the right
   * data type in the payload as well as in original entity data. Without that, we would end up with diffs
   * including all Date properties, as we would be comparing Date object with timestamp.
   */
  override processDateProperty(value: unknown): string | number | Date {
    if (value instanceof Date) {
      return +value;
    }

    return value as number;
  }

  override getIndexName(tableName: string, columns: string[], type: 'index' | 'unique' | 'foreign' | 'primary' | 'sequence'): string {
    if (type === 'primary') {
      return this.getDefaultPrimaryName(tableName, columns);
    }

    return super.getIndexName(tableName, columns, type);
  }

  override getDefaultPrimaryName(tableName: string, columns: string[]): string {
    return 'primary';
  }

  override supportsDownMigrations(): boolean {
    return false;
  }

  override getFullTextWhereClause(): string {
    return `:column: match :query`;
  }

}
