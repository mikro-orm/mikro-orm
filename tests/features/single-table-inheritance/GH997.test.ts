import { Entity, MikroORM, PrimaryKey, Property, OneToMany, ManyToOne, Collection, QueryOrder } from '@mikro-orm/core';
import { SqliteDriver } from '@mikro-orm/sqlite';

abstract class Base {

  @PrimaryKey()
  id!: number;

}

@Entity({
  discriminatorColumn: 'type',
  abstract: true,
  tableName: 'parent_table',
})
class Parent extends Base {

  @Property()
  type!: string;

  @OneToMany(() => Relation1, e => e.parent)
  qaInfo = new Collection<Relation1>(this);

}

@Entity()
class Relation1 extends Base {

  @ManyToOne()
  parent!: Parent;

}

@Entity({ discriminatorValue: 'Child1' })
class Child1 extends Parent {

  @OneToMany(() => Child1Specific, e => e.child1)
  rel = new Collection<Child1Specific>(this);

}

@Entity()
class Child1Specific extends Base {

  @ManyToOne()
  child1!: Child1;

}

@Entity({ discriminatorValue: 'Child2' })
class Child2 extends Parent {
}

describe('GH issue 997', () => {

  let orm: MikroORM<SqliteDriver>;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [Base, Relation1, Child1Specific, Parent, Child1, Child2],
      dbName: ':memory:',
      driver: SqliteDriver,
    });
    await orm.schema.refreshDatabase();
  });

  afterAll(async () => {
    await orm.close(true);
  });

  test(`GH issue 997`, async () => {
    // for unknown reason, for me setup from GH845.test.ts doesn't work,
    // complaining about default values for table columns,
    // so i tweaked it until it works
    const c1 = new Child1();
    const c2 = new Child2();
    await orm.em.persistAndFlush([c1, c2]);
    orm.em.clear();

    const [ci1, ci2]: [Child1, Child2] = await orm.em.find(Parent, {}, {
      populate: ['qaInfo.parent', 'rel'] as never,
      orderBy: { type: QueryOrder.ASC },
    }) as any;

    ci1.rel.add(new Child1Specific());
    ci1.rel.add(new Child1Specific());
    ci1.rel.add(new Child1Specific());
    ci2.qaInfo.add(new Relation1());
    ci2.qaInfo.add(new Relation1());
    ci2.qaInfo.add(new Relation1());

    await orm.em.persistAndFlush([ci1, ci2]);
    orm.em.clear();

    const results = await orm.em.createQueryBuilder(Parent)
      .offset(0)
      .limit(10)
      .orderBy({ type: QueryOrder.ASC })
      .getResult();

    const parents = await orm.em.populate(results as Child1[], ['qaInfo.parent', 'rel']);

    expect(parents[0]).toBeInstanceOf(Child1);
    expect(parents[0].type).toBe('Child1');
    expect(parents[0].qaInfo.length).toBe(0);
    expect((parents[0] as Child1).rel.length).toBe(3);
    expect((parents[0] as Child1).rel[0]).toBeInstanceOf(Child1Specific);
    expect((parents[0] as Child1).rel[0].child1).toBeInstanceOf(Child1);
    expect((parents[0] as Child1).rel[1]).toBeInstanceOf(Child1Specific);
    expect((parents[0] as Child1).rel[1].child1).toBeInstanceOf(Child1);
    expect((parents[0] as Child1).rel[2]).toBeInstanceOf(Child1Specific);
    expect((parents[0] as Child1).rel[2].child1).toBeInstanceOf(Child1);
    expect(parents[1]).toBeInstanceOf(Child2);
    expect(parents[1].type).toBe('Child2');
    expect(parents[1].qaInfo.length).toBe(3);
    expect(parents[1].qaInfo[0]).toBeInstanceOf(Relation1);
    expect(parents[1].qaInfo[0].parent).toBeInstanceOf(Child2);
    expect(parents[1].qaInfo[1]).toBeInstanceOf(Relation1);
    expect(parents[1].qaInfo[1].parent).toBeInstanceOf(Child2);
    expect(parents[1].qaInfo[2]).toBeInstanceOf(Relation1);
    expect(parents[1].qaInfo[2].parent).toBeInstanceOf(Child2);
    expect((parents[1] as Child1).rel).toBeUndefined();
  });

  test(`GH issue 2356`, async () => {
    await expect(orm.schema.getCreateSchemaSQL({ wrap: false })).resolves.toMatchSnapshot();
  });

});
