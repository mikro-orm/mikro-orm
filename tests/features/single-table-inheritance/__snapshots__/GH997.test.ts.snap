// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GH issue 997 > GH issue 2356 1`] = `
"create table \`parent_table\` (\`id\` integer not null primary key autoincrement, \`type\` text not null);
create index \`parent_table_type_index\` on \`parent_table\` (\`type\`);

create table \`child1specific\` (\`id\` integer not null primary key autoincrement, \`child1_id\` integer not null, constraint \`child1specific_child1_id_foreign\` foreign key (\`child1_id\`) references \`parent_table\` (\`id\`) on update cascade);
create index \`child1specific_child1_id_index\` on \`child1specific\` (\`child1_id\`);

create table \`relation1\` (\`id\` integer not null primary key autoincrement, \`parent_id\` integer not null, constraint \`relation1_parent_id_foreign\` foreign key (\`parent_id\`) references \`parent_table\` (\`id\`) on update cascade);
create index \`relation1_parent_id_index\` on \`relation1\` (\`parent_id\`);
"
`;
